python/cuml/test/conftest.py

```diff
@@ -1,19 +1,130 @@
+                                               entry.reprfileloc.lineno)
+
+                    bad_cuml_array_loc.add(
+                        (true_path, entry.reprfileloc.message))
+
+                    break
+
+
+# Closing hook to display the file/line numbers at the end of the test
+def pytest_unconfigure(config):
+    def split_exists(filename: str) -> bool:
+        strip_colon = filename[:filename.rfind(":")]
+        return os.path.exists(strip_colon)
+
+    if (len(bad_cuml_array_loc) > 0):
+
+        print("Incorrect CumlArray uses in class derived from \
+            cuml.common.base.Base:")
+
+        prefix = ""
+
+        # Depending on where pytest was launched from, it may need to append
+        # "python"
+        if (not os.path.basename(os.path.abspath(
+                os.curdir)).endswith("python")):
+            prefix = "python"
+
+        for location, message in bad_cuml_array_loc:
+
+            combined_path = os.path.abspath(location)
+
+            # Try appending prefix if that file doesnt exist
+            if (not split_exists(combined_path)):
+                combined_path = os.path.abspath(os.path.join(prefix, location))
+
+                # If that still doesnt exist, just use the original
+                if (not split_exists(combined_path)):
+                    combined_path = location
+
+            print("{} {}".format(combined_path, message))
+
+        print("See https://github.com/rapidsai/cuml/issues/2456#issuecomment-666106406 \
+            for more information on naming conventions")
+
+
+# This fixture will monkeypatch cuml.common.base.Base to check for incorrect
+# uses of CumlArray.
+@pytest.fixture(autouse=True)
+def fail_on_bad_cuml_array_name(monkeypatch):
+
+    from cuml.common import CumlArray
+    from cuml.common.base import Base
+    from cuml.common.input_utils import get_supported_input_type
+
+    def patched__setattr__(self, name, value):
+
+        supported_type = get_supported_input_type(value)
+
+        if (supported_type == CumlArray):
+            assert name.startswith("_"), "Invalid CumlArray Use! CumlArray \
+                attributes need a leading underscore. Attribute: '{}' In: {}" \
+                    .format(name, self.__repr__())
+        elif (supported_type is not None):
+            # Is this an estimated property? If so, should always be CumlArray
+            assert not name.endswith("_"), "Invalid Estimated Array-Like \
+                Attribute! Estimated attributes should always be CumlArray. \
+                Attribute: '{}' In: {}".format(name, self.__repr__())
+            assert not name.startswith("_"), "Invalid Public Array-Like \
+                Attribute! Public array-like attributes should always be \
+                CumlArray. Attribute: '{}' In: {}".format(
+                name, self.__repr__())
+
+        return super(Base, self).__setattr__(name, value)
+
+    # Monkeypatch CumlArray.__setattr__ to assert array attributes have a
```

Comment:
```suggestion
    # Monkeypatch Base.__setattr__ to assert array attributes have a
```
this is monkey patching the Base class and not the CumlArray class, no?

Reply:
You are correct, this doesnt do anything to `CumlArray`. The first step is just to detect and fix incorrect uses. Additional improvements to `CumlArray` and caching are in another branch.